Implement gravatar email without public address

iref opened the pull request:
This PR implements  #586. 

I am quite new to elixir and hexpm project, so it is quite possible that some things can be implemented in better way. 

Emails table migration sets user's primary email as her Gravatar email. Is that correct? I know that at the moment public email is used, but based on db schema multiple public emails can exist per users, so it's harder to write generic migration using public email. I would welcome any suggestion how to solve that.

If you have any questions, please don't hesitate to ask. :smile:

ericmj commented on lib/hexpm/accounts/users.ex:
```diff
@@ -189,6 +194,52 @@ defmodule Hexpm.Accounts.Users do
+        |> audit(audit_data, "email.gravatar", {new_gravatar, nil})
+    end
+  end
+
```
This code is very similar for public, primary and gravatar emails. Maybe we can look at reducing the code duplication?

iref replied:
Sounds like a good idea. I'll try to simplify it.

ericmj commented on priv/repo/migrations/20171201142544_set_gravatar_on_emails.exs:
```diff
@@ -0,0 +1,14 @@
+
+  def down do
+  end
+end
```
Let's combine this into a single migration.

ericmj commented on test/hexpm/web/views/dashboard_view_test.exs:
```diff
@@ -0,0 +1,42 @@
+
+    assert email == "none"
+  end
+end
```
👏

iref commented:
I commited refactoring changes that remove most of the duplication in email flags. I switched ordering of operations in `primary_email`, so we can compose `User.disable_password_reset` with flag switching transaction. Does it matter in this case?

I also fixed views that used public email instead of gravatar.

ericmj commented on lib/hexpm/web/views/dashboard_view.ex:
```diff
@@ -35,6 +35,19 @@ defmodule Hexpm.Web.DashboardView do
+      |> Enum.filter(&(&1.verified))
+      |> Enum.map(&{&1.email, &1.email})
+
+    [{"Don't use gravatar", "none"}] ++ emails
```
Maybe "Don't show an avatar"?

iref replied:
Maybe. Seems like better description of empty option.

ericmj commented on priv/repo/migrations/20171201141936_add_gravatar_to_emails.exs:
```diff
@@ -0,0 +1,20 @@
+    alter table(:emails) do
+      add :gravatar, :boolean, default: false, null: false
+    end
+    execute """
```
One whitespace line between.

ericmj commented on priv/repo/migrations/20171201141936_add_gravatar_to_emails.exs:
```diff
@@ -0,0 +1,21 @@
+    execute """
+      UPDATE emails
+      SET gravatar = true
+      WHERE "primary"
```
It was the public email we used for gravatar before.

ericmj commented on test/hexpm/web/controllers/dashboard_controller_test.exs:
```diff
@@ -273,6 +290,55 @@ defmodule Hexpm.Web.DashboardControllerTest do
+
+  test "unknown email cannot be gravatar email", c do
+    email = Enum.find(c.user.emails, &(&1.email == "eric@mail.com"))
+    Hexpm.Accounts.Email.toggle_flag(email, :gravatar, true) |> Hexpm.Repo.update
```
`Hexpm.Repo.update!`

iref commented:
No problem! Thank you for merging. It's been fun problem to work on and to get to know hex.pm project.
